lib/crash_reporter.js

```diff
@@ -0,0 +1,191 @@
+var fs = require('fs');
+var logs = require('./logs');
+var packageJson = require('../package.json');
+var path = require('path');
+var request = require('request');
+
+// the value of the crash reporter preference
+// the value has to one of 'on' or 'off'
+var CRASH_REPORTER_PREFERENCE = 'crash.reporter.preference';
+var SUBMIT_CRASH_URL = 'http://tessel-error-reporter.appspot.com/crashes/submit';
+
+// override for testing
+// SUBMIT_CRASH_URL = 'http://localhost:8080/crashes/submit';
+
+var preferenceFile = path.join(__dirname, '__preferences.json');
+
+var Preferences = function() {
+
+};
+
+Preferences.prototype.read = function(key, defaultValue) {
+  return this.readPreferences()
+    .then(function(contents) {
+      if (contents) {
+        return contents[key] || defaultValue;
+      } else {
+        return defaultValue;
+      }
+    })
+    .catch(function(error) {
+      logs.err('Error reading preference', key, error);
+      return defaultValue;
+    });
+};
+
+Preferences.prototype.write = function(key, value) {
+  // read existing preferences and write
+  var that = this;
+  var promise = new Promise(function(resolve, reject) {
+    that.readPreferences()
+      .then(function(contents) {
+        contents = contents || {};
+        contents[key] = value;
+        fs.writeFile(preferenceFile, JSON.stringify(contents), function(error) {
+          if (error) {
+            logs.err('Error writing preference', key, value);
+            reject(error);
+          } else {
+            resolve();
+          }
+        });
+      })
+      .catch(function(error) {
+        reject(error);
+      });
+  });
+  return promise;
+};
```

Comment:
`__preferences.json` is just a way for us to serialize persistent state. So once the user decides that he no longer want to submit crash reports at all, it stays that way.